pallets/liquid-staking-v2/src/types.rs

```diff
@@ -0,0 +1,131 @@
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct PoolLedger<Balance> {
+    pub total_unstake_amount: Balance,
+    pub total_stake_amount: Balance,
+    pub operation_type: Option<StakingOperationType>,
+}
+
+impl<Balance> PoolLedger<Balance>
+where
+    Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default,
+{
+    pub fn op_after_new_era(&self) -> (StakingOperationType, Balance) {
+        if self.total_stake_amount > self.total_unstake_amount {
+            (
+                StakingOperationType::Bond,
+                self.total_stake_amount - self.total_unstake_amount,
+            )
+        } else if self.total_stake_amount < self.total_unstake_amount {
+            (
+                StakingOperationType::Unbond,
+                self.total_unstake_amount - self.total_stake_amount,
+            )
+        } else {
+            (StakingOperationType::Matching, 0u32.into())
+        }
+    }
+}
+
+/// The single user's stake/unsatke amount in each era
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct UserLedger<Balance> {
+    /// The token amount that user unstake during this era, will be calculated
+    /// by exchangerate and xToken amount
+    pub total_unstake_amount: Balance,
+    /// The token amount that user stake during this era, this amount is equal
+    /// to what the user input.
+    pub total_stake_amount: Balance,
+    /// The token amount that user have alreay claimed before the lock period,
+    /// this will happen because, in matching pool total_unstake_amount and
+    /// total_stake_amount can match each other
+    pub claimed_unstake_amount: Balance,
+    /// The token amount that user have alreay claimed before the lock period,
+    pub claimed_stake_amount: Balance,
+    /// To confirm that before lock period, user can only claim once because of
+    /// the matching.
+    pub claimed_matching: bool,
+}
+
+// (claim_unstake_amount_each_era,claim_stake_amount_each_era)
+pub type WithdrawalAmount<Balance> = (Balance, Balance);
+
+impl<Balance> UserLedger<Balance>
+where
+    Balance: AtLeast32BitUnsigned
+        + FullCodec
+        + Copy
+        + MaybeSerializeDeserialize
+        + Debug
+        + Default
+        + FixedPointOperand,
+{
+    pub fn remaining_withdrawal_limit(&self) -> WithdrawalAmount<Balance> {
+        (
+            self.total_unstake_amount
+                .saturating_sub(self.claimed_unstake_amount),
+            self.total_stake_amount
+                .saturating_sub(self.claimed_stake_amount),
+        )
+    }
+
+    // after matching mechanism，for bond operation, user who unstake can get all amount directly
+    // and user who stake only get the matching part
+    pub fn instant_withdrawal_by_bond(
+        &self,
+        pool: &PoolLedger<Balance>,
+    ) -> WithdrawalAmount<Balance> {
+        (
+            self.total_unstake_amount,
```

Comment:
here if we image in one era we only have one person stake, if exchangeRate = 2

Alice stakes 101KSM and unstakes 50xKSM, then alice is supposed to get:

100KSM & 50xKSM from the calculations but does it make sense?